cpp/ql/src/Likely Bugs/RedundantNullCheckParam.ql

```diff
@@ -0,0 +1,56 @@
+/**
+ * @name Redundant null check or missing null check
+ * @description Checking a parameter for nullness in one path,
+ *              and not in another is likely to be a sign that either
+ *              the check can be removed, or added in the other case.
+ * @kind problem
+ * @id cpp/redundant-null-check-param
+ * @problem.severity recommendation
+ * @tags reliability
+ *       security
+ *       external/cwe/cwe-476
+ */
+
+import cpp
+
+predicate blockDominates(Block check, Block access) {
+  check.getLocation().getStartLine() <= access.getLocation().getStartLine() and
+  check.getLocation().getEndLine() >= access.getLocation().getEndLine()
+}
+
+predicate isCheckedInstruction(VariableAccess unchecked, VariableAccess checked) {
+  checked =
+    any(VariableAccess va |
+      va.getTarget() = unchecked.getTarget()
+  ) and
+//Simple test if the first access in this code path is dereferenced
+  not dereferenced(checked) and
+  blockDominates(checked.getEnclosingBlock(), unchecked.getEnclosingBlock())
```

Comment:
```suggestion
  bbDominates(checked.getBasicBlock(), unchecked.getBasicBlock())
```

I'm guessing you're interested in `BasicBlock` (from the control-flow graph) rather than `Block` (from the tree of statements). A `Block` is a `{ ...; ...; }` statement.

If so, you can delete the location-based `blockDominates` predicate.

Reply:
indeed, thanks

Reply:
In facts, it seems the basic block logic leads to false positives with condition like `if (X && ptr == NULL)`

Reply:
We'll have to investigate that if we ever want to promote this query out of `experimental`, but we can leave it as is for now.